feat: add native Node Inspector profiling to meteor profile

italojs opened the pull request:
## description
This PR introduces the creation of the .cpuprofiling fileS inside [Profile function](https://github.com/meteor/meteor/blob/f8a7e448eed50a51bebad6500a64592a9d95136b/tools/tool-env/profile.ts#L292)

Each time we call Profile we will be able to inspect that function execution and name that with the Profile's name + METEOR_INSPECT_CONTEXT env

## Usage
Current usage in [meteor performance ](https://github.com/italojs/performance/tree/monitor-bundler) monitor-bundler branch:
``` shell
METEOR_INSPECT_OUTPUT="/Users/italojs/dev/meteor/performance/profiles" \
METEOR_INSPECT="bundler.bundle" \
METEOR_CHECKOUT_PATH="/Users/italojs/dev/meteor/meteor" \
METEOR_BUNDLE_SIZE=true \ 
sh '/Users/italojs/dev/meteor/performance/scripts/monitor-bundler.sh' '/Users/italojs/dev/meteor/performance/apps/app-3.2'
```

Output files: 
``` shell
372K    ./bundler.bundle-rebuild-client-2025-03-13T15-50-38-301Z.cpuprofile
832K    ./bundler.bundle-rebuild-server-2025-03-13T15-50-55-747Z.cpuprofile
1.1M    ./bundler.bundle-rebuild-client-2025-03-13T15-50-36-542Z.cpuprofile
1.5M    ./bundler.bundle-rebuild-server-2025-03-13T15-50-53-283Z.cpuprofile
5.3M    ./bundler.bundle-rebuild-client-2025-03-13T15-50-31-331Z.cpuprofile
5.4M    ./bundler.bundle-rebuild-server-2025-03-13T15-50-48-255Z.cpuprofile
7.0M    ./bundler.bundle-cache-start-2025-03-13T15-50-18-255Z.cpuprofile
 23M    ./bundler.bundle-cold-start-2025-03-13T15-50-03-055Z.cpuprofile
 60M    ./bundler.bundle-visualize-bundle-2025-03-13T15-51-21-410Z.cpuprofile
104M    .
```

Then, we can open it on https://discoveryjs.github.io/cpupro/
![Screenshot 2025-03-13 at 12 57 00](https://github.com/user-attachments/assets/a5159823-21df-47e9-b5e4-1be13cd149b5)

###  new environment variables

| Variable                        | Description                                                                                               | Default value                |
|---------------------------------|-----------------------------------------------------------------------------------------------------------|------------------------------|
| `METEOR_INSPECT`                 | Enables or disables the Inspector. Can be a comma-separated list of event names to inspect. **The values must match the ones currently used inside the `Profile` wrapper.** If set, the inspector will be active. | `undefined` (disabled)        |
| `METEOR_INSPECT_CONTEXT`         | Sets the name or context for this inspector run, used to identify the profiling output.                   | `'unknown'`                  |
| `METEOR_INSPECT_OUTPUT`          | Directory path where profiling output files will be stored.                                               | `<project folder>/profiling` |
| `METEOR_INSPECT_INTERVAL`        | Sampling interval in milliseconds. Lower values give more detail but consume more memory.                  | `undefined` (optional)        |
| `METEOR_INSPECT_MAX_SIZE`        | Maximum profile file size in MB.                                                                          | `2000`                       |

METEOR_INSPECT_INTERVAL and METEOR_INSPECT_MAX_SIZE is important to handle OOM errors

Note: this PR depends of https://github.com/meteor/performance/pull/11

nachocodoner commented on tools/tool-env/profile.ts:
```diff
@@ -238,66 +296,326 @@ export function Profile<
+          process.once(signal, () => {
+            if (profileInfo.isActive && !profileInfo.isCompleted) {
+              stopInspectorProfiling(name, true).catch(err => {
+                process.stdout.write(`[PROFILING_SIGNAL] Erro ao finalizar profiling: ${err}`);
```
Eu consigo entender essa língua latina parecida, mas provavelmente usamos inglês aqui. 🤣

italojs replied:
lol, i forgot to remove this one, sorry

nachocodoner commented on tools/tool-env/profile.ts:
```diff
@@ -238,66 +296,326 @@ export function Profile<
+    
+    // Adjust the heap limit to avoid OOM
+    if (process.env.NODE_OPTIONS && !process.env.NODE_OPTIONS.includes('--max-old-space-size')) {
+      process.stdout.write('[PROFILING_START] WARN: Recommended to set NODE_OPTIONS="--max-old-space-size=4096" to avoid OOM\n');
```
Have you tried debugging this line with the tool? I think we use TOOL_NODE_FLAGS with the Meteor CLI, while NODE_OPTIONS probably doesn’t work.

italojs replied:
Nice catch!
I fixed my OOM issue using `METEOR_INSPECT_INTERVAL` and `METEOR_INSPECT_MAX_SIZE`. 
I'll double-check it the `--max-old-space-size` flag

nachocodoner commented:
The PR description is well-documented. 👏

It would be a good idea to add this to the docs, possibly where we explain METEOR_PROFILE or in a new section.

Grubba27 commented on package.json:
```diff
@@ -81,5 +81,8 @@
       "curly": "off",
       "sort-imports": "off"
     }
+  },
+  "dependencies": {
+    "simdjson": "^0.9.2"
   }
```
I think this should be a devDependency

Grubba27 commented on tools/isobuild/bundler.js:
```diff
@@ -171,6 +171,7 @@ import { CORDOVA_PLATFORM_VERSIONS } from '../cordova';
 import { gzipSync } from "zlib";
 import { PackageRegistry } from "../../packages/core-runtime/package-registry.js";
 import { optimisticLStatOrNull } from '../fs/optimistic';
+var simdjson = require('simdjson');
```
I think it would be fine to use import here(`import simdjson from "simdjson"`)

italojs commented:
I'm sorry, guys, I converted this to a draft; it's still in progress.

italojs committed:
feat: native profile inspector
- Enhance profiling capabilities with inspector support and update documentation.
- Added .cursorignore and .cursorrules to .gitignore.
- Updated PERFORMANCE.md and README.md with detailed instructions for using inspector profiling, including environment variable configurations and output handling.
- Improved profiling function to manage inspector sessions and handle large profile sizes effectively.

italojs committed:
log: adding –max_old_space_size warn

italojs committed:
refactor: streamline profiling termination handling and cleanup

- Consolidated termination logic for profiling into a single function to reduce redundancy.
- Removed unnecessary comments and code related to heap limit warnings.
- Improved clarity in handling synchronous results during profiling completion.
- Enhanced error logging for profiling termination events.
